src/integrators/ptracer.cpp

```diff
@@ -0,0 +1,245 @@
+#include <random>
+
+#include <enoki/morton.h>
+#include <mitsuba/core/fwd.h>
+#include <mitsuba/core/plugin.h>
+#include <mitsuba/core/progress.h>
+#include <mitsuba/core/properties.h>
+#include <mitsuba/core/ray.h>
+#include <mitsuba/render/bsdf.h>
+#include <mitsuba/render/emitter.h>
+#include <mitsuba/render/integrator.h>
+#include <mitsuba/render/records.h>
+#include <mitsuba/render/sampler.h>
+
+#include <fstream>
+
+// #define MTS_DEBUG_PTRACER_PATHS "/tmp/ptracer.obj"
+#if defined(MTS_DEBUG_PTRACER_PATHS)
+#include <fstream>
+namespace {
+static size_t export_counter = 0;
+} // namespace
+#endif
+
+NAMESPACE_BEGIN(mitsuba)
+
+/**
+ * Traces rays from the light source and attempts to connect them to the sensor
+ * at each bounce.
+ */
+template <typename Float, typename Spectrum>
+class ParticleTracerIntegrator : public LightTracerIntegrator<Float, Spectrum> {
+public:
+    MTS_IMPORT_BASE(LightTracerIntegrator, m_max_depth, m_rr_depth)
+    MTS_IMPORT_TYPES(Scene, Sensor, Sampler, ImageBlock, Medium, Emitter,
+                     EmitterPtr, BSDF, BSDFPtr)
+
+    explicit ParticleTracerIntegrator(const Properties &props) : Base(props) {}
+
+    std::pair<Ray3f, Spectrum> prepare_ray(const Scene *scene,
+                                           const Sensor *sensor,
+                                           Sampler *sampler,
+                                           Mask active) const override {
+        Float time = sensor->shutter_open();
+        if (sensor->shutter_open_time() > 0)
+            time += sampler->next_1d(active) * sensor->shutter_open_time();
+
+        /* ---------------------- Sample ray from emitter ------------------- */
+        // Prepare random samples.
+        auto wavelength_sample = sampler->next_1d(active);
+        auto direction_sample  = sampler->next_2d(active);
+        auto position_sample   = sampler->next_2d(active);
+        // Sample one ray from an emitter in the scene.
+        auto [ray, ray_weight, unused] = scene->sample_emitter_ray(
+            time, wavelength_sample, direction_sample, position_sample, active);
+        return std::make_pair(ray, ray_weight);
+    }
+
+    std::pair<Spectrum, Float>
+    trace_light_ray(Ray3f ray, const Scene *scene, const Sensor *sensor,
+                    Sampler *sampler, SurfaceInteraction3f &si,
+                    Spectrum throughput, int depth, ImageBlock *block,
```

Comment:
`depth` should likely not be a parameter (or it's at least not clear to me why it is.)